src/search/merge_and_shrink/transition_system.cc

```diff
@@ -328,63 +396,67 @@ void TransitionSystem::apply_abstraction(
         utils::g_log << tag() << "initial state pruned; task unsolvable" << endl;
     }
 
-    assert(are_transitions_sorted_unique());
-    assert(in_sync_with_label_equivalence_relation());
+    assert(is_valid());
 }
 
 void TransitionSystem::apply_label_reduction(
     const vector<pair<int, vector<int>>> &label_mapping,
     bool only_equivalent_labels) {
-    assert(are_transitions_sorted_unique());
-    assert(in_sync_with_label_equivalence_relation());
-
     /*
       We iterate over the given label mapping, treating every new label and
       the reduced old labels separately. We further distinguish the case
       where we know that the reduced labels are all from the same equivalence
       class from the case where we may combine arbitrary labels.
 
       The case where only equivalent labels are combined is simple: remove all
-      old labels from the label group and add the new one.
+      old labels from the local label they belong to and add the new one to
+      the local label instead.
 
       The other case is more involved: again remove all old labels from their
-      groups, and the groups themselves if they become empty. Also collect
-      the transitions of all reduced labels. Add a new group for every new
-      label and assign the collected transitions to this group. Recompute the
-      cost of all groups and compute locally equivalent labels.
-
-      NOTE: Previously, this latter case was computed in a more incremental
-      fashion: Rather than recomputing cost of all groups, we only recomputed
-      cost for groups from which we actually removed labels (hence temporarily
-      storing these affected groups). Furthermore, rather than computing
-      locally equivalent labels from scratch, we did not per default add a new
-      group for every label, but checked for an existing equivalent label
-      group. In issue539, it turned out that this incremental fashion of
-      computation does not accelerate the computation.
+      local labels, and the local labels themselves if do not represent any
+      global labels anymore. Also collect the transitions of all reduced
+      labels. Add a new local label for every new label and assign the
+      collected transitions to it. Recompute the cost of all affected local
+      labels and re-compute locally equivalent labels.
```

Comment:
I first misunderstood this comment in regards to what happens for equivalent labels. I didn't understand that we add the new global label to the same label group that the old global labels were in, which in turn means that the affected local label cannot become empty. I'm not sure if the comment can be improved to make this clearer or if it's just because I'm not familiar enough with the details of M&S ;).